tools/triage/ticket_search/jira_attachment_downloader.py

```diff
@@ -0,0 +1,75 @@
+import time
+import jira
+import os
+from .file_path_recursor import FilePathRecursor
+from .action_extract_archive import ActionExtractArchive
+from tools.triage.utils.extraction import GzipFileExtractor
+from tools.triage.utils.extraction import ShutilExtractor
+from tools.triage.ticket_search import settings
+from tools.triage.utils.filesystem import FileSystemHelpers
+from tools.triage.ticket_search.cli.workflow_data import WorkflowData
+class JiraAttachmentDownloader:
+    """Downloads Jira attachments for a ticket and extracts their contents"""
+    def __init__(self, jira_client:"jira", workflow_data:"WorkflowData"):
+        """Constructor"""
+        self.extraction_recursor = FilePathRecursor(
+            on_file_action_class=ActionExtractArchive(
+                extractors=[
+                    ShutilExtractor(),
+                    GzipFileExtractor()
+                ]
+            ),recurse_after_action=True
+        )
+        self.jira_client = jira_client
+        self.workflow_data = workflow_data
+
+    def _handle_error(self, message, is_fatal=True):
+        """Handles an error if it occurs"""
+        print(message)
+        if is_fatal:
+            raise JiraAttachmentDownloaderException(message)
+
+    def download_attachments_for_ticket(self, issue):
+        download_start_time = datetime.datetime.now()
+        print(f"Downloading {issue.key} as it is not present in the cache {download_start_time}")
+        time.sleep(settings.JIRA_ATTACHMENT_PER_TICKET_THROTTLE_SECONDS)
+        try:
+            destination_path = os.path.join(FileSystemHelpers.get_current_user_directory(), "triage-tools-tickets", issue.key)
+            if os.path.exists(destination_path):
+                # If the directory has already been downloaded from the issue then skip it
+                print(f"Skipping download for {issue.key} as this has already been downloaded. {datetime.datetime.now()}")
+                return
+            FileSystemHelpers.mkdir(destination_path)
+            if len(issue.fields.attachment) > 0:
+                for attachment in issue.fields.attachment:
+                    # We occasionally get an issue here if the server is busy, does not appear to be possible to detect this easily.
+                    # Downloaded content will be the HTML for a Jira error page but I don't think we have access to response code.
+                    # Official docs do not appear to show any error handling.
+                    # https://jira.readthedocs.io/examples.html#attachments
+                    # Might be better to use http API to obtain download URL and use the requests library so that we can obtain status codes etc.
+                    content = attachment.get() 
+                    filename = os.path.join(destination_path, attachment.filename)
+                    FileSystemHelpers.write_file_content(filename, content)
+                    time.sleep(settings.JIRA_ATTACHMENT_PER_ITEM_THROTTLE_SECONDS)
```

Comment:
Probably would leave this to blow up and backoff only then, as we adding unnecessary delay

Reply:
Will do in another PR